src/bootstrap/suggest.rs

```diff
@@ -0,0 +1,142 @@
+        .output()
+        .expect("failed to execute `git`");
+
+    let modified_files =
+        String::from_utf8(modified_files.stdout).expect("`git` gave non-utf8 commit?!");
+
+    let modified_files: Vec<_> = modified_files.split_terminator('\n').collect();
+
+    println!("=== MODIFIED FILES ===");
+    for file in &modified_files {
+        println!("{}", file);
+    }
+    println!("======================\n");
+
+    let mut suggestions = Vec::new();
+    for sug in SUGGESTIONS {
+        let glob = Pattern::new(sug.0).expect("Found invalid glob pattern!");
+
+        for file in &modified_files {
+            if glob.matches(file) {
+                suggestions.extend_from_slice(sug.1);
+            }
+        }
+    }
+
+    suggestions.sort();
+    suggestions.dedup();
+
+    println!("=== SUGGESTIONS ===");
+    for sug in &suggestions {
+        println!("{}", sug);
+    }
+    println!("===================");
+
+    if run {
+        for _sug in suggestions {
+            // HELP!: how should I use `rustbuild`'s internal plumbing to achieve this?
+        }
+    } else {
+        println!("help: consider using the `--run` flag to automatically run suggested tests");
+    }
+}
+
+#[derive(Clone, PartialOrd, Ord, PartialEq, Eq)]
+struct Suggestion {
+    subcommand: &'static str,
+    stage: Option<u8>,
+    paths: &'static [&'static str],
+}
+
+impl Display for Suggestion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "x.py {} ", self.subcommand)?;
+
+        if let Some(stage) = self.stage {
+            write!(f, "--stage {} ", stage)?;
+        }
+
+        for path in self.paths {
+            write!(f, "{} ", path)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Suggestion {
+    pub const fn new(
+        subcommand: &'static str,
+        stage: Option<u8>,
+        paths: &'static [&'static str],
+    ) -> Self {
+        Self { subcommand, stage, paths }
+    }
+}
+
+macro_rules! sug {
+    ($sub:expr) => {
+        Suggestion::new($sub, None, &[])
+    };
+
+    ($sub:expr, $paths:expr) => {
+        Suggestion::new($sub, None, $paths)
+    };
+
+    ($sub:expr, $stage:expr, $paths:expr) => {
+        Suggestion::new($sub, Some($stage), &$paths)
+    };
+}
+
+// FIXME: perhaps this could use `std::lazy` when it is stablizied
+macro_rules! suggestions {
+    ($( $glob:expr => [ $( $suggestion:expr ),* ] ),*) => {
+        const SUGGESTIONS: &[(&str, &[Suggestion])] = &[ $( ($glob, &[ $($suggestion),* ]) ),*];
+    }
+}
+
+suggestions! {
+    "*.md" => [
+        sug!("test", 0, ["tidy", "linkchecker"])
+    ],
+
+    "library/std/src/lib.rs" => [
+        sug!("test", 0, ["tidy", "library/std"])
+    ],
+
+    "compiler/rustc_data_structures/*" => [
+        sug!("check"),
+        sug!("test", 0, ["tidy", "compiler/rustc_data_structures"]),
+        sug!("test", 1, ["src/test/ui", "src/test/run-make"])
+    ],
+
+    "src/librustdoc/src/lib.rs" => [
```

Comment:
```suggestion
    "src/librustdoc/*" => [
```

Reply:
Obviously these suggestions are really just examples and have to be built on over time.